### ISSUE TITLE
ES6 feature: Iteration using for of

### ISSUE THREAD

**Issue by ariya** [jquery/esprima#1047]
### Iteration Statements

<pre>
IterationStatement[Yield, Return] :
  do Statement[?Yield, ?Return] while ( Expression[In, ?Yield] ) ;opt
  while ( Expression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( [lookahead ∉ {let [}] Expression[?Yield]opt ; Expression[In, ?Yield]opt ; Expression[In, ?Yield]opt ) Statement[?Yield, ?Return]
  for ( var VariableDeclarationList[?Yield] ; Expression[In, ?Yield]opt ; Expression[In, ?Yield]opt ) Statement[?Yield, ?Return]
  for ( LexicalDeclaration[?Yield] Expression[In, ?Yield]opt ; Expression[In, ?Yield]opt ) Statement[?Yield, ?Return]
  for ( [lookahead ∉ {let [}] LeftHandSideExpression[?Yield] in Expression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( var ForBinding[?Yield] in Expression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( ForDeclaration[?Yield] in Expression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( [lookahead ≠ let ] LeftHandSideExpression[?Yield] of AssignmentExpression[In, ?Yield] ) Statement [?Yield, ?Return]
  for ( var ForBinding[?Yield] of AssignmentExpression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( ForDeclaration[?Yield] of AssignmentExpression[In, ?Yield] ) Statement[?Yield, ?Return]

ForDeclaration[Yield] :
  LetOrConst ForBinding[?Yield]

ForBinding[Yield] :
  BindingIdentifier[?Yield]
  BindingPattern[?Yield]
</pre>

### for-of Statements

```
IterationStatement :
  for ( LeftHandSideExpression in Expression ) Statement
  for ( LeftHandSideExpression of AssignmentExpression ) Statement
```

## Spec:

https://people.mozilla.org/~jorendorff/es6-draft.html#sec-for-in-and-for-of-statements

Relevant AST interfaces (from https://github.com/estree/estree):

```js
interface ForOfStatement <: ForInStatement {
    type: "ForOfStatement";
}
```

## Early Errors:

The following early errors should be detected:

```js
for (const of 42);     // Unexpected number
for (this of that);    // Invalid left-hand side in for-loop

// Init is forbidden
for (var x = 1 of y);
for (let x = 1 of y);
for (const x = 1 of y);
```

## Related Tasks:

- [ ] Error message. V8 currently just says "Invalid left-hand side in for-loop"

**Comment by gotwarlost** [comment]
Any way you guys can prioritize this feature to happen sooner? This is supported by node, esprima#harmony and istanbul#harmony. I really want to get rid of the istanbul harmony branch and get everyone to use the mainline.

**Comment by mikesherov** [comment]
@gotwarlost I'll see if there's any way to squeeze this into 2.2. Thanks for the report.

**Comment by gotwarlost** [comment]
Thanks!

**Comment by ariya** [comment]
@gotwarlost This for-of iteration is now supported in the freshly released 2.4.0!

**Comment by gotwarlost** [comment]
Thanks! I'll start using it, er, _soon_ - not sure when :)
